cirq/protocols/json_test.py

```diff
@@ -0,0 +1,446 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import inspect
+
+import io
+import os
+
+import pytest
+
+import numpy as np
+import sympy
+
+import cirq
+import cirq.protocols
+
+
+def assert_roundtrip(obj, text_should_be=None):
+    buffer = io.StringIO()
+    cirq.protocols.to_json(obj, buffer)
+
+    if text_should_be is not None:
+        buffer.seek(0)
+        text = buffer.read()
+        assert text == text_should_be
+
+    buffer.seek(0)
+    obj2 = cirq.protocols.read_json(buffer)
+    if isinstance(obj, np.ndarray):
+        np.testing.assert_equal(obj, obj2)
+    else:
+        assert obj == obj2
+
+
+def test_line_qubit_roundtrip():
+    q1 = cirq.LineQubit(12)
+    assert_roundtrip(q1,
+                     text_should_be="""{
+  "cirq_type": "LineQubit",
+  "x": 12
+}""")
+
+
+def test_gridqubit_roundtrip():
+    q = cirq.GridQubit(15, 18)
+    assert_roundtrip(q,
+                     text_should_be="""{
+  "cirq_type": "GridQubit",
+  "row": 15,
+  "col": 18
+}""")
+
+
+def test_op_roundtrip():
+    q = cirq.LineQubit(5)
+    op1 = cirq.Rx(.123).on(q)
+    assert_roundtrip(op1,
+                     text_should_be="""{
+  "cirq_type": "GateOperation",
+  "gate": {
+    "cirq_type": "XPowGate",
+    "exponent": 0.03915211600060625,
+    "global_shift": -0.5
+  },
+  "qubits": [
+    {
+      "cirq_type": "LineQubit",
+      "x": 5
+    }
+  ]
+}""")
+
+
+def test_op_roundtrip_filename(tmpdir):
+    filename = f'{tmpdir}/op.json'
+    q = cirq.LineQubit(5)
+    op1 = cirq.Rx(.123).on(q)
+    cirq.to_json(op1, filename)
+    assert os.path.exists(filename)
+    op2 = cirq.read_json(filename)
+    assert op1 == op2
+
+
+def test_fail_to_resolve():
+    buffer = io.StringIO()
+    buffer.write("""
+    {
+      "cirq_type": "MyCustomClass",
+      "data": [1, 2, 3]
+    }
+    """)
+    buffer.seek(0)
+
+    with pytest.raises(ValueError) as e:
+        cirq.read_json(buffer)
+    assert e.match("Could not resolve type 'MyCustomClass' "
+                   "during deserialization")
+
+
+QUBITS = cirq.LineQubit.range(5)
+Q0, Q1, Q2, Q3, Q4 = QUBITS
+
+TEST_OBJECTS = {
+    'CCNOT':
+    cirq.CCNOT,
```

Comment:
yapf did this to you?

Reply:
yes. are you asking about how it put the keys and values on separate lines but with the same indentation?